### ISSUE TITLE
Python `store.dump()` serialization causes Segfault with very large Stores

### ISSUE THREAD

**Issue by ashleysommer** [oxigraph/oxigraph#1078]
**Describe the bug**
Parsing a large RDF file into an in-memory PyOxigraph store, then using `dump()` to serialize it back out again, causes a segfault to be thrown from module `pyoxigraph.cpython-311-x86_64-linux-gnu.so` after `dump()` is completed, causing Python to crash.

A "large graph" is relative in the RDF world, in this case I'm loading a 650MB NQuads file, when parsed the store contains approx 2.8M quads, in 2 named graphs.

Cannot reproduce it using a 1MB NQuads file. I don't know where the tipping point is, somewhere between 1MB and 650MB.

**To Reproduce**
Using PyOxigraph v0.4.2 on Python 3.11

```python
from pyoxigraph import Store as oxigraph_Store, RdfFormat
    
def reproduce():
    my_file = Path("/home/user/Downloads/very_large_file.nq")
    inner = oxigraph_Store()
    with open(my_file, "rb") as f:
        inner.bulk_load(f, format=RdfFormat.N_QUADS)

    out_file = str(my_file).replace(".nq", ".test.nq")
    print(f"Writing out file {out_file}")
    with open(out_file, "wb") as f_out:
        inner.dump(out_file, format=RdfFormat.N_QUADS)
    print("Done writing file.")

if __name__ == "__main__":
    reproduce()
    print("Exiting")
```
Output:
```
Writing out file /home/user/Downloads/my_large_file.test.nq
Done writing file.
Exiting

Process finished with exit code 139 (interrupted by signal 11:SIGSEGV)
```
I see in my kernel log:
```log
Nov 05 10:12:26 hostname kernel: python[2131662]: segfault at 7ffe918c4ff8 ip 0000761faef843fe sp 00007ffe918c5000 error 6 in pyoxigraph.cpython-311-x86_64-linux-gnu.so[761faee9b000+a71000] likely on CPU 4 (core 8, socket 0)
Nov 05 10:12:26 hostname kernel: Code: 5f 5d ff 25 dc bd c2 00 48 83 c4 28 5b 41 5c 41 5d 41 5e 41 5f 5d c3 0f 1f 44 00 00 41 56 53 50 48 89 fb 4c 8b 37 49 8d 7e 10 <e8> 0d 1d ff ff 41 83 be b8 00 00 00 03 75 17 49 83 be c8 00 00 00
```
Important to note:
* the produced output file is created correctly with the right content.
* The segfault is _after_ "Exiting" is printed, indicating the segfault occurs when python is finished executing and is cleaning up memory 
* I've tried dump() with `N_QUADS` and `TRIG`, both produce the segfault, I haven't tested others.
* The format used in parsing doesn't matter, tried all of them
* `bulk_load()` vs `load()` doesn't matter, I tried both.
* `File` object vs `str` path in `dump()` produces the same segfault

**Comment by ashleysommer** [comment]
Update:
I tested with a 151MB NQuad file with 573,943 quads, and I still get the Segfault.

**Comment by Tpt** [comment]
Thank you for the bug report! I am going to investigate that

**Comment by ashleysommer** [comment]
Another data point.
Loading that same 65MB nquads file as above, then adding a new named graph to the dataset with around 30 more triples, then serializing it back on out _does_ produce the segfault.

This could also have something to do with memory pressure. I'm running this on a system with 32GB ram, but 24GB is already used by other processes. Loading the very large 650MB file does push the memory availability to the limit.

I'll see if I can reproduce it on a different system with different memory availability.

**Comment by Tpt** [comment]
I don't think it's a memory pressure problem. I managed to reproduce the bug while still having more than half of my RAM free. My bet is that it's some issue at the end of the process (some bad destructor/memory deallocation...)

**Comment by Tpt** [comment]
By the way, you can directly load data from a file (it skips Python I/O abstraction entirely and you open the output file in Python even if you directly write to it.

Simplified code:
```python
from pyoxigraph import Store as oxigraph_Store, RdfFormat
    
def reproduce():
    my_file = Path("/home/user/Downloads/very_large_file.nq")
    inner = oxigraph_Store()
    inner.bulk_load(path=my_file, format=RdfFormat.N_QUADS)

    out_file = str(my_file).replace(".nq", ".test.nq")
    print(f"Writing out file {out_file}")
    inner.dump(out_file, format=RdfFormat.N_QUADS)
    print("Done writing file.")

if __name__ == "__main__":
    reproduce()
    print("Exiting")
```

This code is still failing so bug is likely not related to my wrapper on top of Python IO objects.

**Comment by Tpt** [comment]
It is indeed a destructor-related issue causing a stack overflow. I am preparing a fix.

**Comment by Tpt** [comment]
Fix: #1082

**Comment by ashleysommer** [comment]
> By the way, you can directly load data from a file

I know. I did mention that I tried that in my initial comment.
> File object vs str path in dump() produces the same segfault

The reason I'm using a Python File IO object in my code is because my nquads files are actually gzipped nquads.
My code looks like this:

```python
    my_file = Path("/home/user/very_large_file.nq.gz")
    f = GzipFile(my_file, "rb")
    try:
        inner.bulk_load(f, format=RdfFormat.N_QUADS)
    finally:
        f.close()
    ...
    out_file = str(my_file).replace(".nq.gz", ".test.nq.gz")
    print(f"Writing out file {out_file}")
    f_out = GzipFile(out_file, mode="wb")
    try:
        inner.dump(f_out, format=RdfFormat.N_QUADS)
    finally:
        f_out.close()
```

I know the oxigraph CLI tool does support parsing and serializing gzipped files, I've used that before, but using gzipped files this way though Python directly to the store load/dump is not supported. (I suppose that is a feature request?)

**Comment by Tpt** [comment]
Thank you for the details!

> I suppose that is a feature request?

Not yet, buf feel free to open it

**Comment by Tpt** [comment]
Should be fixed in v0.4.3. Please reopen if it's not the case. Thank you again for the bug report

**Comment by ashleysommer** [comment]
@Tpt  Yes, I've been testing since v0.4.3 came out, and no more segfaults.
